[READY] Automatically load a compilation database if found

puremourning opened the pull request:
Alternative implementation to : https://github.com/Valloric/ycmd/pull/679

See #679 for explanation and rationale.

Fixes: https://github.com/Valloric/ycmd/issues/489
Fixes: https://github.com/Valloric/YouCompleteMe/issues/2310 (sort of)
Fixes: https://github.com/Valloric/ycmd/issues/26
Fixes: https://github.com/Valloric/YouCompleteMe/issues/1981
Fixes: https://github.com/Valloric/YouCompleteMe/issues/1623 (sort of)
Fixes: https://github.com/Valloric/YouCompleteMe/issues/1622
Partly implements: https://github.com/Valloric/YouCompleteMe/issues/927
Fixes: https://github.com/Valloric/YouCompleteMe/issues/664
Fixes: https://github.com/Valloric/YouCompleteMe/issues/487
Fixes (discussion on): https://github.com/Valloric/YouCompleteMe/issues/174

This implementation is almost identical to the default-extra-conf version, except:

- the code lives in `flags.py`
- it re-uses `INCLUDE_FLAGS` from `flags.py`
- it provides the directory of the compilation database in the `clang_completer` debug info instead of the name of the default extra conf module.

On reflection, I think I might prefer this implementation.

For now, the tests are identical to the other PR (you can now see why I added that set of tests!)

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/680)
<!-- Reviewable:end -->

Valloric commented:
@r4nt I remember you wanting to see this implemented in ycmd. Curious what your thoughts are; check out both this PR and the other implementation (linked in PR description).

Valloric commented:
Gave a light review pass. I'm leaning towards this approach as well now that I've seen both. The other approach is a default extra conf file, but the we load it by default and the user would be unwise to change that file directly... at that point we might as well just have this default logic hard-coded into ycmd, because it effectively is.

As usual, thanks for the PR! :)

---

Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[cpp/ycm/ClangCompleter/CompilationDatabase.h, line 57 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_Mg2GjYFe8w4yqm-zD:-K_Mg2GjYFe8w4yqm-zE:b-6qgpxx) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/cpp/ycm/ClangCompleter/CompilationDatabase.h#L57)):*
> ```objc
>     const boost::python::object &path_to_file );
> 
>   const char * GetDatabaseDirectory()
> ```

I'm pretty sure you can just make this return an `std::string` and Boost.Python will know what to do with it. 

Let's avoid returning raw pointers if possible.

---

*[ycmd/completers/cpp/flags.py, line 96 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_MgLAFaFt6hPSKB4PA:-K_MgLAFaFt6hPSKB4PB:b-vulwm8) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/ycmd/completers/cpp/flags.py#L96)):*
> ```Python
>     # We cache the database for any given source directory
>     self.compilation_database_dir_map = dict()
>     # Sometimes we don't actually know what the flags to use are. Rather than
> ```

Empty line before comment block?

---

*[ycmd/completers/cpp/flags.py, line 232 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_MgqQ3eWAeA30O2Yct:-K_MgqQ3eWAeA30O2Ycu:byxwf0r) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/ycmd/completers/cpp/flags.py#L232)):*
> ```Python
>   # database already for that path, or if a compile_commands.json file exists in
>   # that directory.
>   def FindCompilationDatabase( self, wd ):
> ```

Let's expand `wd` to a longer name.

---

*[ycmd/completers/cpp/flags.py, line 235 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_MgxOWfccb1CQWHUu0:-K_MgxOWfccb1CQWHUu1:boilh7i) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/ycmd/completers/cpp/flags.py#L235)):*
> ```Python
>     # Find all the ancestor directories of the supplied directory
>     for folder in PathsToAllParentFolders( wd ):
>       # Did we already cache a database for this path?
> ```

Maybe cut down on some of these obvious comments now that this isn't in a user-facing file. :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680#-:-K_MhMld0lnzsImRmMEy:br2pljj)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 0 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[cpp/ycm/ClangCompleter/CompilationDatabase.h, line 57 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_Mg2GjYFe8w4yqm-zD:-K_MjEDH4WMWhxQXx2rp:b6ky3uj) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/cpp/ycm/ClangCompleter/CompilationDatabase.h#L57)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

I'm pretty sure you can just make this return an `std::string` and Boost.Python will know what to do with it. 

Let's avoid returning raw pointers if possible.
</blockquote></details>

Old habit. Trying to avoid a string copy (`std::string&` and `const std::string&` don't work). But truth is I didn't check whether the underlying implementation doesn't copy the string anyway, so meh, changed to `std::string` return.

---

*[ycmd/completers/cpp/flags.py, line 96 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_MgLAFaFt6hPSKB4PA:-K_MjyU_89VSEZDA1T6y:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/ycmd/completers/cpp/flags.py#L96)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Empty line before comment block?
</blockquote></details>

Done.

---

*[ycmd/completers/cpp/flags.py, line 232 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_MgqQ3eWAeA30O2Yct:-K_Mk-D11hN21hEhYJWN:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/ycmd/completers/cpp/flags.py#L232)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Let's expand `wd` to a longer name.
</blockquote></details>

Done.

---

*[ycmd/completers/cpp/flags.py, line 235 at r1](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_MgxOWfccb1CQWHUu0:-K_MkpzkLNl__T2GJaJb:b613ko0) ([raw file](https://github.com/valloric/ycmd/blob/affc3a0438b4d8bf38a97ca4f6e0d6de5dc8933b/ycmd/completers/cpp/flags.py#L235)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Maybe cut down on some of these obvious comments now that this isn't in a user-facing file. :)
</blockquote></details>

Yep, for this revision i literally copy/pasted the whole code block :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

vheon commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/cpp/clang_completer.py, line 37 at r3](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_QA8uAtZJoXQ4ABaBy:-K_QA8uAtZJoXQ4ABaBz:b21a449) ([raw file](https://github.com/valloric/ycmd/blob/91558c0bd585f3c140d164bc64748ae6e5c441b1/ycmd/completers/cpp/clang_completer.py#L37)):*
> ```Python
> from ycmd.completers.completer import Completer
> from ycmd.completers.completer_utils import GetIncludeStatementValue
> from ycmd.completers.cpp.flags import ( Flags, PrepareFlagsForClang )
> ```

why the parenthesis here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/cpp/clang_completer.py, line 37 at r3](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_QA8uAtZJoXQ4ABaBy:-K_Qns7U3WbkzbR0t7iB:b-60uwgc) ([raw file](https://github.com/valloric/ycmd/blob/91558c0bd585f3c140d164bc64748ae6e5c441b1/ycmd/completers/cpp/clang_completer.py#L37)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

why the parenthesis here?
</blockquote></details>

I think because i changed this then changed it back. Fixed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

r4nt commented:
Yep, agree that built-in is probably better. Very happy to see projects that use the compilation db work just out of the box. Thanks so much for this.
If I had one wish, it would be longest common suffix matching for .h/.cpp paths, but that might be too brittle :)

puremourning commented:
> If I had one wish, it would be longest common suffix matching for .h/.cpp paths, but that might be too brittle

Interesting. Thanks! We were chatting about this last night. Our plan is go out something simple live then improve the heuristics based on feedback and/or inspiration. We considered more caches and some path manipulations for header files (the latter being the key goal). 

Do you have experience with longest common suffix? Do you know that it works well? If so, that could be our first feedback/inspiration point. :)

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice stuff! Some minor inline comments, otherwise great.

---

Review status: 0 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/cpp/flags.py, line 97 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyBmMDp1Y-4SpFr0V:-K_VyBmMDp1Y-4SpFr0W:b-840ppg) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L97)):*
> ```Python
> 
>     # We cache the compilation database for any given source directory
>     self.compilation_database_dir_map = dict()
> ```

Since this is python, I'd like to see docs stating what the type of key and value are.

---

*[ycmd/completers/cpp/flags.py, line 104 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyHiTJG_B2YW0ZQyD:-K_VyHiTJG_B2YW0ZQyE:bpy1o8p) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L104)):*
> ```Python
>     # percentage of cases and allow new files (which are not yet in the
>     # compilation database) to receive at least some flags
>     self.file_directory_heuristic_map = dict()
> ```

Same as above; I'd like to see docs stating what the type of key and value are.

God, I hate dynamic languages...

---

*[ycmd/tests/clang/__init__.py, line 104 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_W5lPaGF_xFZTt8R6K:-K_W5lPaGF_xFZTt8R6L:b-3w2w76) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/tests/clang/__init__.py#L104)):*
> ```Python
> 
> @contextlib.contextmanager
> def TemporaryProject( tmp_dir, compile_commands ):
> ```

"Project" seems to be used here to imply a `compile_command.json` connection. That's not very obvious to me. Maybe "ClangProject"? Something better than "project" which is too ambiguous. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680#-:-K_W6RTfu1JfD7khaDXj:b1sjhd0)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Thanks for the review!

---

Review status: 0 of 8 files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/completers/cpp/flags.py, line 97 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyBmMDp1Y-4SpFr0V:-K_WEuFTtc1jHWvbHl0D:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L97)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Since this is python, I'd like to see docs stating what the type of key and value are.
</blockquote></details>

Done.

---

*[ycmd/completers/cpp/flags.py, line 104 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyHiTJG_B2YW0ZQyD:-K_WEurwMQ2uhAUuNsbC:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L104)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Same as above; I'd like to see docs stating what the type of key and value are.

God, I hate dynamic languages...
</blockquote></details>

Done.

---

*[ycmd/tests/clang/__init__.py, line 104 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_W5lPaGF_xFZTt8R6K:-K_WFxVu9zxb01pcG8q8:b-9svjn8) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/tests/clang/__init__.py#L104)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

"Project" seems to be used here to imply a `compile_command.json` connection. That's not very obvious to me. Maybe "ClangProject"? Something better than "project" which is too ambiguous. 
</blockquote></details>

Done. ClangProject seems ok.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680#-:-K_WFzZzzhNUdzD-YufO:b-9qdoz3)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 2 of 6 files at r1, 1 of 3 files at r2, 1 of 5 files at r4, 1 of 5 files at r5.
Review status: 4 of 8 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/cpp/clang_completer.py, line 388 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_Vz2Swa-qG8NS8phfR:-K_Vz2SxdEZVKGjP1vzV:bpj2469) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/clang_completer.py#L388)):*
> ```Python
>       database = self._flags.FindCompilationDatabase(
>           os.path.dirname( filename ) )
>       if database is None or not database.DatabaseSuccessfullyLoaded():
> ```

can't we write this as `if not database`?

---

*[ycmd/completers/cpp/flags.py, line 104 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WRbFthVTkYQfxtyMr:-K_WRbFuMCi9a7P9fN0N:b8gl2pn) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L104)):*
> ```Python
>     # Sometimes we don't actually know what the flags to use are. Rather than
>     # returning no flags, if we've previously found flags for a file in a
>     # particular directory, return them. The will probably work in a high
> ```

Just a nitpick but is "The" right here?

---

*[ycmd/completers/cpp/flags.py, line 198 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WSNQTbLFzZHNWPoii:-K_WSNQTbLFzZHNWPoij:b-exwxxj) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L198)):*
> ```Python
> 
>     database = self.FindCompilationDatabase( file_dir )
>     if database is None or not database.DatabaseSuccessfullyLoaded():
> ```

We call this function in two places and in both places we do this check. Couldn't we `raise` directly in `FindCompilationDatabase` and adapt the other call to `try: catch` this? I'm not sure that is better but since this check is done twice I leaning to think that it deserves at least a name (even though is a simple condition)?

---

*[ycmd/completers/cpp/flags.py, line 205 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WT8l-Lz4myt47JlB6:-K_WT8l0ph-6SQ8c1GcY:b-ck3bgz) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L205)):*
> ```Python
>                                                    file_extension )
> 
>     if compilation_info is None:
> ```

I think is more Pythonic to write it as `if not compilation_info`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

micbou commented:
Reviewed 3 of 6 files at r1, 3 of 3 files at r2, 2 of 2 files at r3, 5 of 5 files at r4, 4 of 5 files at r5.
Review status: all files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[cpp/ycm/ClangCompleter/CompilationDatabase.h, line 58 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_OxCzZwQAMW3nejQ5X:-K_OxCz_txXQL1Zj2_xf:b-f13dnn) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/cpp/ycm/ClangCompleter/CompilationDatabase.h#L58)):*
> ```objc
> 
>   std::string GetDatabaseDirectory()
>   {
> ```

Isn't the open bracket supposed to be on the first line according to our coding style?

---

*[ycmd/completers/cpp/clang_completer.py, line 402 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_Oxofu_qL5FPHz3E4G:-K_Oxofu_qL5FPHz3E4H:b-a61z1q) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/completers/cpp/clang_completer.py#L402)):*
> ```Python
>                  '  No configuration file found\n'
>                  '  No compilation database file found\n' )
>       else:
> ```

This `else:` is not needed since we are returning in the first branch.

---

*[ycmd/completers/cpp/clang_completer.py, line 391 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_TYWDubVEMPaZcTKtD:-K_TYWDubVEMPaZcTKtE:b7eknor) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/clang_completer.py#L391)):*
> ```Python
>         return ( 'C-family completer debug information:\n'
>                  '  No configuration file found\n'
>                  '  No compilation database found\n' )
> ```

Last `\n` should be removed.

---

*[ycmd/completers/cpp/clang_completer.py, line 397 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WUlPaT4mnqhevzVPJ:-K_WUlPbhWEWqyXei8XH:bw24i6u) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/clang_completer.py#L397)):*
> ```Python
>                  '  Using compilation database from: {0}\n'
>                  '  Flags: {1}'.format( database.database_directory,
>                                         list( flags ) ) )
> ```

It's not obvious why `flags`is always defined in this case and the one below. I think we should move the `flags` assignment in its own `try/catch` block like this:
```python
try:
  extra_conf = extra_conf_store.ModuleFileForSourceFile( filename )
except UnknownExtraConf as error:
  return ...

try:
  flags = self._FlagsForRequest( request_data ) or []
except NoExtraConfDetected:
  return ...

database = self._flags.compilation_database_dir_map.get( 
    os.path.dirname( filename ) )

if database:
  return ... flags ...
return ... flags ...
```
This way, it is clear that `flags` is always defined.

---

*[ycmd/completers/cpp/flags.py, line 186 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_P1reTJzmNf8edlTVn:-K_P1reUpv81Ec7qdXCY:bkbsnxx) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/completers/cpp/flags.py#L186)):*
> ```Python
> 
>   def Clear( self ):
>     self.flags_for_file.clear()
> ```

I think this function should also clear the `compilation_database_dir_map` and `file_directory_heuristic_map` dictionaries.

---

*[ycmd/completers/cpp/flags.py, line 191 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_OytiA58qzByk4be0K:-K_OytiA58qzByk4be0L:by9ifnv) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/completers/cpp/flags.py#L191)):*
> ```Python
>   def _GetFlagsFromCompilationDatabase( self, file_name ):
>     file_dir = os.path.dirname( file_name )
>     ( file_root, file_extension ) = os.path.splitext( file_name )
> ```

Parentheses are not required.

---

*[ycmd/completers/cpp/flags.py, line 248 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WcIzfrCGIK_vqu2wV:-K_WcIzfrCGIK_vqu2wW:b-vcl197) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L248)):*
> ```Python
>     # Nothing was found. No compilation flags are available.
>     # Note: we cache the fact that none was found for this folder to speed up
>     # subsequent searches..
> ```

One dot can be removed.

---

*[ycmd/tests/clang/__init__.py, line 133 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_TZAIJe7dntXcoiUeR:-K_TZAIJe7dntXcoiUeS:b-ai34kw) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/tests/clang/__init__.py#L133)):*
> ```Python
>     yield
>   finally:
>     os.unlink( path )
> ```

We use `os.remove` in other places so we should also use it here to be consistent ([`os.remove` and `os.unlink` are identicals](https://docs.python.org/2/library/os.html#os.unlink)).

---

*[ycmd/tests/clang/flags_test.py, line 513 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_P-lZOR9zZuCvcdWj4:-K_P-lZPkaUMYNwLJvgs:bug6i0m) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/tests/clang/flags_test.py#L513)):*
> ```Python
> 
> 
> def _MakeRelativePathsInFlagsAvsoluteTest( test ):
> ```

Typo: `Avsolute` instead of `Absolute`.

---

*[ycmd/tests/clang/flags_test.py, line 513 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_TZg9AJr-jWTNrMn4l:-K_TZg9AJr-jWTNrMn4m:b32behq) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/tests/clang/flags_test.py#L513)):*
> ```Python
> 
> 
> def CompilationDatabase_MakeRelativePathsInFlagsAbsolute_test():
> ```

This test is not directly related to compilation databases so I would not prefix its name with `CompilationDatabase`. Same comment for the tests below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 3 of 8 files reviewed at latest revision, 17 unresolved discussions.

---

*[cpp/ycm/ClangCompleter/CompilationDatabase.h, line 58 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_OxCzZwQAMW3nejQ5X:-K_Wh7MwwL4QkqXeoY0N:b-ghzh4t) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/cpp/ycm/ClangCompleter/CompilationDatabase.h#L58)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Isn't the open bracket supposed to be on the first line according to our coding style?
</blockquote></details>

it is indeed.

---

*[ycmd/completers/cpp/clang_completer.py, line 402 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_Oxofu_qL5FPHz3E4G:-K_WhFt54D8B3nHq2kOM:b-b1sx4z) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/completers/cpp/clang_completer.py#L402)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

This `else:` is not needed since we are returning in the first branch.
</blockquote></details>

Done.

Removed by refactor of exceptions.

---

*[ycmd/completers/cpp/clang_completer.py, line 388 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_Vz2Swa-qG8NS8phfR:-K_Wh8weDhGQ0ABMDQ6S:b-alwgpu) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/clang_completer.py#L388)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

can't we write this as `if not database`?
</blockquote></details>

yeah, i just hate dynamic languages. None behaves strangely.

---

*[ycmd/completers/cpp/clang_completer.py, line 391 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_TYWDubVEMPaZcTKtD:-K_WhEjHc2CRSR83OGKA:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/clang_completer.py#L391)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Last `\n` should be removed.
</blockquote></details>

Done.

---

*[ycmd/completers/cpp/clang_completer.py, line 397 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WUlPaT4mnqhevzVPJ:-K_WhIWuJnhWN9RGBwGX:b-pz7p6k) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/clang_completer.py#L397)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

It's not obvious why `flags`is always defined in this case and the one below. I think we should move the `flags` assignment in its own `try/catch` block like this:
```python
try:
  extra_conf = extra_conf_store.ModuleFileForSourceFile( filename )
except UnknownExtraConf as error:
  return ...

try:
  flags = self._FlagsForRequest( request_data ) or []
except NoExtraConfDetected:
  return ...

database = self._flags.compilation_database_dir_map.get( 
    os.path.dirname( filename ) )

if database:
  return ... flags ...
return ... flags ...
```
This way, it is clear that `flags` is always defined.
</blockquote></details>

I've changed the control flow here to make it more obvious and added a few comments.

---

*[ycmd/completers/cpp/flags.py, line 186 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_P1reTJzmNf8edlTVn:-K_WmTbyauZk16wHgTkK:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/completers/cpp/flags.py#L186)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

I think this function should also clear the `compilation_database_dir_map` and `file_directory_heuristic_map` dictionaries.
</blockquote></details>

Done.

---

*[ycmd/completers/cpp/flags.py, line 191 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_OytiA58qzByk4be0K:-K_WmUH5NfW7S1vsvxpM:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/completers/cpp/flags.py#L191)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Parentheses are not required.
</blockquote></details>

Done.

---

*[ycmd/completers/cpp/flags.py, line 104 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WRbFthVTkYQfxtyMr:-K_WmSFvA9tNgo6xyW0J:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L104)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

Just a nitpick but is "The" right here?
</blockquote></details>

Done.

---

*[ycmd/completers/cpp/flags.py, line 198 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WSNQTbLFzZHNWPoii:-K_WX1lRcQ6oFWNg-4jt:btsvvhl) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L198)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

We call this function in two places and in both places we do this check. Couldn't we `raise` directly in `FindCompilationDatabase` and adapt the other call to `try: catch` this? I'm not sure that is better but since this check is done twice I leaning to think that it deserves at least a name (even though is a simple condition)?
</blockquote></details>

Yeah, I have it this way in another change. It is better to have `FindCompilatoinDatabase` raise `NoCompilationDatabase` This required a bit of refactoring.

---

*[ycmd/completers/cpp/flags.py, line 205 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WT8l-Lz4myt47JlB6:-K_WmXfahkgbuCMYkDOT:b-p7jzdz) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L205)):*
<details><summary><i>Previously, vheon (Andrea Cedraro) wrote…</i></summary><blockquote>

I think is more Pythonic to write it as `if not compilation_info`?
</blockquote></details>

grr. done :)

---

*[ycmd/completers/cpp/flags.py, line 248 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_WcIzfrCGIK_vqu2wV:-K_Wm_MSEr7WJQrIg0se:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/completers/cpp/flags.py#L248)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

One dot can be removed.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Review status: 3 of 8 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/cpp/flags.py, line 97 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyBmMDp1Y-4SpFr0V:-K_XLjbtCHQmARSf45_f:bjxbdkl) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L97)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Maybe use "key" instead of "LHS" and "value" instead of "RHS"? Seems more obvious. 

---

*[ycmd/completers/cpp/flags.py, line 104 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyHiTJG_B2YW0ZQyD:-K_XLzZ-2LBuYxDPsWOu:b-y2gvxr) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L104)):*
<details><summary><i>Previously, puremourning (Ben Jackson) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Same as above. RHS is a needless abbreviation, just use "value".

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Reviewed 2 of 6 files at r1, 1 of 5 files at r5, 5 of 5 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 7 of 8 files reviewed at latest revision, 9 unresolved discussions.

---

*[ycmd/tests/clang/__init__.py, line 133 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_TZAIJe7dntXcoiUeR:-K_ZcCGc80N7mVc7xhHL:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/tests/clang/__init__.py#L133)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

We use `os.remove` in other places so we should also use it here to be consistent ([`os.remove` and `os.unlink` are identicals](https://docs.python.org/2/library/os.html#os.unlink)).
</blockquote></details>

Done.

---

*[ycmd/tests/clang/flags_test.py, line 513 at r2](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_P-lZOR9zZuCvcdWj4:-K_ZcCr-_-lsfevaJwPY:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/0bc399561300a7d73f17bb821fb83dc8af9671c6/ycmd/tests/clang/flags_test.py#L513)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

Typo: `Avsolute` instead of `Absolute`.
</blockquote></details>

Done.

---

*[ycmd/tests/clang/flags_test.py, line 513 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_TZg9AJr-jWTNrMn4l:-K_ZcDUg8U-NnRZ64FZa:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/tests/clang/flags_test.py#L513)):*
<details><summary><i>Previously, micbou wrote…</i></summary><blockquote>

This test is not directly related to compilation databases so I would not prefix its name with `CompilationDatabase`. Same comment for the tests below.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 6 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/completers/cpp/flags.py, line 97 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyBmMDp1Y-4SpFr0V:-K_Zctr79_kG2itl5f8U:bb6un1j) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L97)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Maybe use "key" instead of "LHS" and "value" instead of "RHS"? Seems more obvious. 
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Review status: 6 of 8 files reviewed at latest revision, 6 unresolved discussions.

---

*[ycmd/completers/cpp/flags.py, line 104 at r4](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_VyHiTJG_B2YW0ZQyD:-K_ZcyDke6whgOJ6SRj0:b-896fix) ([raw file](https://github.com/valloric/ycmd/blob/bf093d32403f61707677d10fd4972d024806fc10/ycmd/completers/cpp/flags.py#L104)):*
<details><summary><i>Previously, Valloric (Val Markovic) wrote…</i></summary><blockquote>

Same as above. RHS is a needless abbreviation, just use "value".
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[ycmd/tests/clang/debug_info_test.py, line 98 at r5](https://reviewable.io:443/reviews/valloric/ycmd/680#-K_bgU-qv9CL-luIeheP:-K_bgU-qv9CL-luIeheQ:bkwaq9u) ([raw file](https://github.com/valloric/ycmd/blob/b868dd265ff652fe131f8914a5139e7564046ba2/ycmd/tests/clang/debug_info_test.py#L98)):*
> ```Python
>                         '  No configuration file found\n'
>                         '  Using compilation database from: {0}\n'
>                         '  Flags: .+-Wall.+'.format( tmp_dir ) ) )
> ```

Oh nice!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680)*
<!-- Sent from Reviewable.io -->

micbou commented:
I tested it on our code base and, to my surprise, it worked great. I only needed to copy the `compile_commands.json` file into the root project. We could even drop our `.ycm_extra_conf.py` file with this PR (or maybe not).

<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r6, 2 of 2 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680#-:-K_g7Ym32MMxJmg5Uv7R:bhirhwi)*
<!-- Sent from Reviewable.io -->

puremourning commented:
I've rebased and just tweaked a few comments which were wrong or contained typos.

Is this ready to merge now? I was thinking we could merge this, merge my documentation updates to YCM and update ycmd submodule. This would release python 3.6 support, a few bug fixes and this change. Major release! :)

---

Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/680#-:-K_tNJ5L4Bv6xv8vJwwS:b-x6p61z)*
<!-- Sent from Reviewable.io -->

Valloric commented:
Merge it! :)

puremourning commented:
@homu r=micbou

homu commented:
:pushpin: Commit 128b508 has been approved by `micbou`

<!-- @homu r=micbou 128b508a220f1e7bff9818295abdcf604466af1f -->

homu commented:
:hourglass: Testing commit 128b508 with merge d19f9c5...

homu commented:
:sunny: Test successful - [status](https://travis-ci.org/Valloric/ycmd/builds/189852399)
